cgi-bin/DW/Controller/Admin/SendMail.pm

```diff
@@ -41,96 +42,112 @@ sub index_controller {
-                  $msgid, $remote->id, time, $account, $sendto, $subject,
-                  $support_req, $message, $teamnotes )
-            or return error_ml( '/admin/sendmail/index.tt.error.sendfailed' );
-
-        # 2. construct the message and send it to the user(s)
-        # (this block adapted from bin/worker/paidstatus)
-        my $send = { from => "$account\@$LJ::DOMAIN",
-                     fromname => $LJ::SITENAME,
-                     subject => $subject,
-                     body => $message,
-                   };
-        my $sent = 0;
-
-        if ( $u && $u->is_community ) {
-            # send an email to every maintainer
-            my $maintus = LJ::load_userids( $u->maintainer_userids );
-            foreach my $maintu ( values %$maintus ) {
-                if ( $send->{to} = $maintu->email_raw ) {
-                    LJ::send_mail( $send );
-                    $sent = 1;
+        my $u;
+
+        if ( $sendto ) {
+            if ( $sendto =~ /,/ ) { # multiple recipients
+                $errors->add( "sendto", ".error.nomulti" );
+            } else {
+                # make sure we're sending to either a valid user or something
+                # that looks reasonably like an email address.
+                if ( $sendto !~ /^[^@]+@[^.]+\./ ) {
+                    # doesn't look like an email address; do a username lookup
+                    $u = LJ::load_user( $sendto );
+                    $errors->add( "sendto", ".error.badrcpt" ) unless defined $u;
+                    $sendto = $u->id;  # log userid instead of username
                 }
             }
-        } elsif ( $u ) {
-            if ( $send->{to} = $u->email_raw ) {
-                LJ::send_mail( $send );
-                $sent = 1;
-            }
-        } else {
-            $send->{to} = $sendto;
-            LJ::send_mail( $send );
-            $sent = 1;
+        } else { # no $sendto
+            $errors->add( "sendto", ".error.norcpt" );
         }
 
-        return error_ml( '/admin/sendmail/index.tt.error.nouseremail' )
-            unless $sent;
+        # at this point, we should have good form data; the form can still
+        # have errors below, but they aren't the fault of the user input
+
+        unless ( $errors->exist ) {
+            # now that we have the data, send the message.
+            # 1. insert data into siteadmin_email_history table
+            my $msgid = LJ::alloc_global_counter('N');
+            my $dbh = LJ::get_db_writer();
+            $dbh->do( "INSERT INTO siteadmin_email_history (msgid, remoteid," .
+                      " time_sent, account, sendto, subject, request, message, " .
+                      " notes) VALUES (?,?,?,?,?,?,?,?,?)", undef,
+                      $msgid, $remote->id, time, $account, $sendto, $subject,
+                      $support_req, $message, $teamnotes )
+                or return error_ml( $scope->( '.error.sendfailed' ) );
```

Comment:
sure. in general I try for error_ml for things the user can't change, and the $errors-> stuff for things we want to show the form again for, so that the user can change something.